src/audio/module_adapter/module/generic.c

```diff
@@ -83,12 +83,6 @@ int module_init(struct processing_module *mod, struct module_interface *interfac
 
 	comp_dbg(dev, "module_init() start");
 
-#if CONFIG_IPC_MAJOR_3
-	if (mod->priv.state == MODULE_INITIALIZED)
-		return 0;
-	if (mod->priv.state > MODULE_INITIALIZED)
-		return -EPERM;
-#endif
```

Comment:
@kv2019i this is safe to remove for IPC3 but as for IPC4, I think i'd like @dbaluta to confirm is it is OK to remove for the cadence codecs

Reply:
@dbaluta , will IPC3 also support cadence codecs?

Reply:
@btian1  yes!

Reply:
so we can't remove those ipc3 specific code? @dbaluta @lyakh @ranj063, since ipc3 need it? from my observation, I did not see those code is must even for IPC3

Reply:
@btian1 we can't remove IPC3 specific code as IMX uses it. 

This patch works fine for imx cadence tests. You just need to do the following changes:

```
diff --git a/src/audio/module_adapter/module/cadence.c b/src/audio/module_adapter/module/cadence.c
index ef912d7e2..dd0654228 100644
--- a/src/audio/module_adapter/module/cadence.c
+++ b/src/audio/module_adapter/module/cadence.c
@@ -353,7 +353,7 @@ static int cadence_codec_init(struct processing_module *mod)
        codec->mpd.init_done = 0;
 
        /* copy the setup config only for the first init */
-       if (codec->state == MODULE_DISABLED && codec->cfg.avail) {
+       if (codec->cfg.avail) {
                setup_cfg = &cd->setup_cfg;
 
                /* allocate memory for set up config */
@@ -852,7 +852,6 @@ cadence_codec_set_configuration(struct processing_module *mod, uint32_t config_i
                                const uint8_t *fragment, size_t fragment_size, uint8_t *response,
                                size_t response_size)
 {
-       struct module_data *md = &mod->priv;
        struct comp_dev *dev = mod->dev;
        int ret;
```

Reply:
@dbaluta yes, I did it last night, if this patch works fine for imx ipc3 cadence codec test, means md->state can be removed for both ipc4 and ipc3, if so, could you add your approval to this PR?

@lgirdwood ack, due to CI does not check ipc3 build, some place change are ignored, I added last night.

This change is not a irrelevant change, it is critical for host performance in #7909